### DISCUSSION TITLE
Questions about OPA integration with MinIO

### DISCUSSION THREAD

**Discussion by olivier-jobert** [minio/minio#16516] [Q&A] [answered]
Hello,

1. According to this [discussion](https://github.com/minio/minio/discussions/12422#discussioncomment-816361), integrating OPA with MinIO does not seems to be a priority for the MinIO team, from what I understand.

Is it to be understood that there will be no further developments around this feature **or** that the integration with OPA will tend to disappear from MinIO releases?

2. I tried to couple OPA with MinIO based on [OPA Quickstart guide](https://github.com/minio/minio/blob/master/docs/iam/opa.md) and it works.

But when I look at OPA logs, I can see that kind of input:

```yaml
"input": {
  "account": "user1",
  "action": "s3:ListAllMyBuckets",
  "bucket": "",
  "claims": {},
  "conditions": {
    "Authorization": [
      "AWS4-HMAC-SHA256 Credential=user1/..."
    ],
    "CurrentTime": [
      "YYYY-MM-DDTHH:MM:SSZ"
    ],
    "EpochTime": [
      "1675071198"
    ],
    "Referer": [
      ""
    ],
    "SecureTransport": [
      "false"
    ],
    "SourceIp": [
      "127.0.0.1"
    ],
    "User-Agent": [
      "MinIO (linux; amd64) minio-go/v7.0.34 mc/RELEASE.2022-08-28T20-08-11Z"
    ],
    "UserAgent": [
      "MinIO (linux; amd64) minio-go/v7.0.34 mc/RELEASE.2022-08-28T20-08-11Z"
    ],
    "X-Amz-Content-Sha256": [
      "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
    ],
    "X-Amz-Date": [
      "YYYYMMDDTHHMMSSZ"
    ],
    "authType": [
      "REST-HEADER"
    ],
    "principaltype": [
      "User"
    ],
    "signatureversion": [
      "AWS4-HMAC-SHA256"
    ],
    "userid": [
      "user1"
    ],
    "username": [
      "user1"
    ],
    "versionid": [
      ""
    ]
  },
  "denyOnly": false,
  "groups": null,
  "object": "",
  "owner": false
}
```

2.1 It looks like when I am  using the MinIO 'root' user owner field is true (`owner: true`), when I'l using any other user (even if it as the `consoleAdmin` policy associated), owner field is false (owner: false`). Does is work as intended?
  
2.2 Same question question about `principaltype` field which value is `principaltype: Account` for 'root' user and `principaltype: User` for any other user.

2.3 Some fields like`input.groups` and `input.objet` seems to be always null or empty, does I missed something?

2.4 Can you confirm me that there is no way to know MinIO policy/group linked to a user in the OPA request made by MinIO? In fact, I would add some specific OPA policy rules depending on MinIO policy/rules.

Thanks for any answer :)

**Comment by harshavardhana** [comment, marked as answer]
We do not have any opa specific plugin what we have is custom

- authn plugin
- authz plugin

So you do what you want on the webhook endpoint, we don't care who is your authn or authz - write your own.

**Comment by harshavardhana** [comment]
> 2.1 It looks like when I am using the MinIO 'root' user owner field is true (owner: true), when I'l using any other user (even if it as the consoleAdmin policy associated), owner field is false (owner: false`). Does is work as intended?

There is only one owner account which is root user, that variable ideally must have been called as `root: true` but that is a historical baggage that we are not planning to break at this moment.

> 2.3 Some fields likeinput.groups and input.objet seems to be always null or empty, does I missed something?

This is only if your user is part of groups, or your user comes from an LDAP, OIDC authority.

object field is empty for listBuckets call, you didn't specify any object in your request.

> 2.2 Same question question about principaltype field which value is principaltype: Account for 'root' user and principaltype: User for any other user.

It is something we borrowed from AWS, this field is not of big consequence, you can use it as you please.

> 2.4 Can you confirm me that there is no way to know MinIO policy/group linked to a user in the OPA request made by MinIO? In fact, I would add some specific OPA policy rules depending on MinIO policy/rules.

If you configure OPA no policy is applied at MinIO layer anymore. You need to manage policies yourself. You can associate them to your users all you want but it won't take any precedence.

OPA (authz) takes precedence over all local  settings.

**Reply by olivier-jobert** [reply]
> > 2.1 It looks like when I am using the MinIO 'root' user owner field is true (owner: true), when I'l using any other user (even if it as the consoleAdmin policy associated), owner field is false (owner: false`). Does is work as intended?
> 
> There is only one owner account which is root user, that variable ideally must have been called as `root: true` but that is a historical baggage that we are not planning to break at this moment.
> 

OK 👍


> > 2.3 Some fields likeinput.groups and input.objet seems to be always null or empty, does I missed something?
> 
> This is only if your user is part of groups, or your user comes from an LDAP, OIDC authority.

I checked again on a dev instance with local users (*without LDAP, OIDC authority*) and even if I create a user then I add it to a MinIO group, `group` field remains `null`:
```shell
sh-4.4# mc alias set myminio http://127.0.0.1:9000 minio password
Added `myminio` successfully.

sh-4.4# mc admin user add myminio user1 userpassword
Added user `user1` successfully.

sh-4.4# mc admin group add myminio group1 user1
Added members {user1} to group group1 successfully.

sh-4.4# mc admin group info myminio group1
Group: group1
Status: enabled
Policy: 
Members: user1

sh-4.4# mc alias set myminio http://127.0.0.1:9000 user1 userpassword
Added `myminio` successfully.

sh-4.4# mc mb myminio/bucket1
Bucket created successfully `myminio/bucket1`.
```

OPA decision log:
```json
{
  "decision_id": "396612f6-7318-463a-a820-b6902dd44008",
  "input": {
    "account": "user1",
    "action": "s3:CreateBucket",
    "bucket": "bucket1",
    "claims": {},
    "conditions": {
	  ...
      "principaltype": [
        "User"
      ],
      "userid": [
        "user1"
      ],
      ...
    },
    "denyOnly": false,
    "groups": null,
    "object": "",
    "owner": false
  },
  ...
}
```

> 
> object field is empty for listBuckets call, you didn't specify any object in your request.
> 

About `object` field, indeed if I make a GetObject, the field is correctly filled 👍.

> > 2.2 Same question question about principaltype field which value is principaltype: Account for 'root' user and principaltype: User for any other user.
> 
> It is something we borrowed from AWS, this field is not of big consequence, you can use it as you please.
> 

OK 👍.

> > 2.4 Can you confirm me that there is no way to know MinIO policy/group linked to a user in the OPA request made by MinIO? In fact, I would add some specific OPA policy rules depending on MinIO policy/rules.
> 
> If you configure OPA no policy is applied at MinIO layer anymore. You need to manage policies yourself. You can associate them to your users all you want but it won't take any precedence.
> 
> OPA (authz) takes precedence over all local settings.

That's very interesting, I had not noticed that the MinIO policies were no longer taken into account when OPA is coupled with MinIO.
So, I can create some specific policies over users/groups only in OPA, but in our use case we need to set up them on groups, so I've to try with an external IdP (*LDAP, OIDC authority*) because at the moment `groups` remains null with local MinIO users (see above).

**Comment by harshavardhana** [comment]
I looked groups claims are added only for `jwt` 

```
        // Add groups claim which could be a list. This will ensure that the claim
        // `jwt:groups` works.
        if grpsVal, ok := claims["groups"]; ok {
                if grpsIs, ok := grpsVal.([]interface{}); ok {
                        grps := []string{}
                        for _, gI := range grpsIs {
                                if g, ok := gI.(string); ok {
                                        grps = append(grps, g)
                                }
                        }
                        if len(grps) > 0 {
                                args["groups"] = grps
                        }
                }
        }
```

**Reply by harshavardhana** [reply]
This can be fixed to add `groups` support for all login types

- MinIO as IDP (self)
- OIDC
- LDAP

**Reply by olivier-jobert** [reply]
Thanks for the clarification!
